performance_manager/lib/l2_headways.py

```diff
@@ -0,0 +1,383 @@
+            sa.literal(None).label("fk_tu_stopped_event"),
+            FullTripEvents.hash,
+            VehiclePositionEvents.direction_id,
+            VehiclePositionEvents.route_id,
+            VehiclePositionEvents.start_date,
+            VehiclePositionEvents.vehicle_id,
+            sa.case(
+                (
+                    StaticStops.parent_station.isnot(None),
+                    StaticStops.parent_station,
+                ),
+                else_=VehiclePositionEvents.stop_id,
+            ).label("parent_station"),
+            VehiclePositionEvents.timestamp_start,
+        )
+        .join(
+            VehiclePositionEvents,
+            FullTripEvents.fk_vp_stopped_event == VehiclePositionEvents.pk_id,
+        )
+        .join(
+            StaticStops,
+            sa.and_(
+                VehiclePositionEvents.fk_static_timestamp
+                == StaticStops.timestamp,
+                VehiclePositionEvents.stop_id == StaticStops.stop_id,
+            ),
+            isouter=True,
+        )
+        .where(
+            sa.and_(
+                FullTripEvents.fk_vp_stopped_event.isnot(None),
+                VehiclePositionEvents.timestamp_start > min_timestamp_start,
+            )
+        )
+    )
+
+    # query for stop event records from trip updates that require headways
+    # processing
+    tu_query = (
+        sa.select(
+            sa.literal(None).label("fk_vp_stopped_event"),
+            FullTripEvents.fk_tu_stopped_event.label("fk_tu_stopped_event"),
+            FullTripEvents.hash,
+            TripUpdateEvents.direction_id,
+            TripUpdateEvents.route_id,
+            TripUpdateEvents.start_date,
+            TripUpdateEvents.vehicle_id,
+            sa.case(
+                (
+                    StaticStops.parent_station.isnot(None),
+                    StaticStops.parent_station,
+                ),
+                else_=TripUpdateEvents.stop_id,
+            ).label("parent_station"),
+            TripUpdateEvents.timestamp_start,
+        )
+        .join(
+            TripUpdateEvents,
+            FullTripEvents.fk_tu_stopped_event == TripUpdateEvents.pk_id,
+        )
+        .join(
+            StaticStops,
+            sa.and_(
+                TripUpdateEvents.fk_static_timestamp == StaticStops.timestamp,
+                TripUpdateEvents.stop_id == StaticStops.stop_id,
+            ),
+            isouter=True,
+        )
+        .where(
+            sa.and_(
+                FullTripEvents.fk_tu_stopped_event.isnot(None),
+                FullTripEvents.fk_vp_stopped_event.is_(None),
+                TripUpdateEvents.timestamp_start > min_timestamp_start,
+            )
+        )
+    )
+
+    # merge stop event results from vehcile positions and trip updates into
+    # one query all records should be unique
+    stop_events_cte = vp_query.union_all(tu_query).cte(name="all_stop_events")
```

Comment:
in an ideal world, do we expect the length of this cte to be the same size as the vp_query and tu_query?

Reply:
Yes, in an ideal world every vehicle position stop event would have a corresponding trip update stop event. But then we also wouldn't need to use the trip update data for this purpose.